### DISCUSSION TITLE
Pass `queryClient` as first argument of `initialDataFn`

### DISCUSSION THREAD

**Discussion by mijamo** [TanStack/query#2764] [Ideas]
My use case:

-  I want to achieve something like https://react-query.tanstack.com/guides/initial-query-data#initial-data-from-cache
- I use SSR and as such I separate my query objects from my components. They live in separate pure JS modules so that I can import the query objects and prefetch them in my server script.
- The current documentation calls `queryClient`, which is NOT an argument of the function but should be optained independantly. In most cases using `useQueryClient` should be enough, but in my case I build the query objects outside my React components, so I cannot use hook.
- As a result I cannot have stable (constant) query objects but I need to have factories that take in the queryClient. It complicates the code and adds some unstability (my initialData function is now recreated everytime my component rerenders for instance)

Maybe there's a better way to handle it but I think passing `queryClient` to the `initialData` makes sense as it is a very common workflow, in the same way as the queryKey is passed to the queryFunction

**Comment by TkDodo** [comment]
> but in my case I build the query objects outside my React components, so I cannot use hook.

`useQuery` calls `useQueryClient` under the hood, so wherever yo do `useQuery`, you can also do `useQueryClient`. If you build the query config outside of react, you'd need to pass the `queryClient` there:

```
const buildQueryConfig = (queryClient) => ...

const MyComponent = () => {
  useQuery(buildQueryConfig(useQueryClient()))
}
```

**Reply by GLabat** [reply]
Jumping into the discussion after reading the last message.

When query configurations are built outside of hooks, queryClient is often required (dependent queries for ex).
As explained above, it is required to pass it.

Inspired by this awesome [article](https://tkdodo.eu/blog/leveraging-the-query-function-context) ;), a queryConfig would look like:

```js
{
  queryKey: keyFactory.detail('someId')
  queryFn: functionRelyingOnQueryFunctionCtx
}
```

Here comes the question:

If `functionRelyingOnQueryFunctionCtx` requires an object given by another query, it makes the configuration a bit more complicated.
Could it be possible that the "ambient" queryClient be part of the QueryFunctionCtx?
This way,  `functionRelyingOnQueryFunctionCtx` could rely on it to retrieve the external object. The logic remains encapsulated and the query configuration simple.

I'm not sure that this is possible but it would be great to have your feedback on it.
Thanks.

**Reply by TkDodo** [reply]
it would be possible to add the queryClient to the queryFunctionContext - but I fail to see which problem that would solve. It would only be available to the `queryFn` - the function that does the actual data fetching - where it wouldn't be of much use...

**Reply by GLabat** [reply]
Hi and thanks for taking the time to answer.

Let me try to explain a bit more.

In our app, we use schemas (JSONSchema-like) to describe our entities.
We setup a helper `function getSchemaQueryConfig(schemaName, queryOptions)` to retrieve the configuration (`queryKey`, `queryFn` and some options) of a query for a given schema.
We also have a hook `useSchema` which uses this helper to build a query config which is then passed to `useQuery`.
This "solves" the problem of retrieving a schema in the "React world".

For the "JS world", a helper `async function getSchema(queryClient, schemaName, queryOptions)` is defined and also relies on `getSchemaQueryConfig` to build the query config and pass it to `fetchQuery`.

With this in place, we do not have to bother anymore about query key, query function, cache options etc. to retrieve a schema, which is great.

We decline this pattern of query configurations to retrieve other entities.
But for the query configurations **for which a schema is required**, this is a bit more complex.
We also want query configurations to be "just" configuration, in the sense that it must only return a configuration, no call, no async stuff, etc.

For ex

```js
function getStuffEntityQueryConfig(stuffOptions, queryOptions) {
return {
    queryKey: stuffKeyFactories.someFactory(stuffOptions),
    queryFn: async queryFunctionCtx => {
      // We want Stuff schema to be retrieved here so that in case of failure,
      // the Stuff query would fail too.
      // This means that the `queryClient` is required here so that we could
      // rely on the cache to get the schema.
      const stuffSchema = await getSchema(queryClient, SCHEMA_STUFF)
      const { extract, some, props, from, schema } = stuffSchema
      return fetchStuffFn(
        queryFunctionCtx,
        { extract, some, props, from, schema }
      )
    },
    ...queryOptions
  }
}
```

As you can see, this pattern requires an instance of the `queryClient`.
One way of having it (this is what we do for now but raised this discussion) is to pass it as argument of `getStuffEntityQueryConfig` => this complexifies a bit the call.

Another way would be to have the `queryClient` in the `queryFunctionCtx` passed to the `queryFn`, which would leave the configuration getter autonomous.

Hope this clarifies a bit.

Thanks.

**Reply by TkDodo** [reply]
hmm, we internally get the `queryClient` via `useQueryClient` ourselves. `useQuery` creates an Observer, which also has the client. Basically, the client doesn't really do anything. We pass it down react context for easy retrieval. If you want to access the client from outside of react, you can also just `export const queryClient = new QueryClient()` where you create it and then import it

**Comment by TkDodo** [comment]
however, composing query functions from other queries of the same client seems useful. for example, you can currently do this via:

```
const queryClient = useQueryClient()
useQuery(key, () => queryClient.getQueryData(something) ?? axios.get(...))
```

if the query function is not inlined, passing the client is necessary, so I agree that having the client injected might be useful in some cases. I'll definitely think about it :)

**Reply by GLabat** [reply]
Cool, can't wait for v4 ;)

Additional question: having a global `queryClient` might be incompatible with several `QueryClientProviders`, each using a different `queryClient`. What do you think?
Having a local `queryClient` inside `QueryFunctionCtx` could help here, it would return the closest one.
But in this case, how would we reach another one up in the tree?
I don't know if nested `QueryClientProviders` is a good practice or not…

**Reply by TkDodo** [reply]
> Additional question: having a global queryClient might be incompatible with several QueryClientProviders, each using a different queryClient. What do you think?

yes, that's why I don't usually recommend it. But I also don't recommend having functions outside of hooks that calculate query options :)
but I agree that injecting the queryClient would likely solve this.

> I don't know if nested QueryClientProviders is a good practice or not…

depends. it creates a new cache as well, unless you copy it over (see: https://tkdodo.eu/blog/testing-react-query#reactqueryconfigprovider). If you want isolated scopes, like for microfrontends, then yes, it is a good way.

**Reply by GLabat** [reply]
> But I also don't recommend having functions outside of hooks that calculate query options :)

Ahhh this triggers a lot of questions on our side 😅.
Does this mean that you would not recommend our "separated query configuration pattern"?

We took inspiration from your patterns and adapted them to our codebase.
The fact that we have to deal with **AND** without React makes us realise that factorisation of query configs would be a great benefice.
Should you see any drawbacks with this approach, please let us know.
And if you are interested and want more information, we could setup something (we sort of plan to write an article about our transition to react-query).

We should also be candidate to help for the modification of QueryFunctionContext above.

Thanks.

**Reply by TkDodo** [reply]
> Does this mean that you would not recommend our "separated query configuration pattern"?

Given the reasons that you've stated (need it for SSR), I would say no, but I don't have much experience with prefetching - and there is also no good / bad pattern here. It's all very nuanced. Very likely you don't want to use the same options on the server than on the client. Like, the callbacks like `onSuccess` / `onError` are not needed, setting up interval refetching is not needed etc. I know this still comes across that way in the docs, but I think most often, you wouldn't even want to re-use the `queryFn`:
- on the server, you might want to call a function that grabs data directly from the database
- while on the client, you will need to call the (rest) endpoint that exposes the data.

I think the SSR example should really be:

```
 // pages/posts.jsx
 import { dehydrate, QueryClient, useQuery } from 'react-query';
 
 export async function getStaticProps() {
   const queryClient = new QueryClient()
 
   await queryClient.prefetchQuery('posts', getPostsFromDB)
 
   return {
     props: {
       dehydratedState: dehydrate(queryClient),
     },
   }
 }
 
 function Posts() {
   // This useQuery could just as well happen in some deeper child to
   // the "Posts"-page, data will be available immediately either way
   const { data } = useQuery('posts', fetchPosts)
 
   // This query was not prefetched on the server and will not start
   // fetching until on the client, both patterns are fine to mix
   const { data: otherData } = useQuery('posts-2', fetchPosts)
 
   // ...
 }
```

I wanted to change this some time ago, but somehow forgot about it :)

> We took inspiration from your patterns and adapted them to our codebase.

what I like to do is have sharable objects / factories for the `queryKeys` - even though I _also_ have them as hooks sometimes (`useQueryKeys()`) because they internally use some id from the url for example. That is needed imo because we have many places where we need to have access to the key, mainly when performing query invalidation.

---

Another good way to share options between `useQuery` and `prefetchQuery`, if necessary, is to use `queryClient.setQueryDefautls`, which sets default options on a per key basis:

```
queryClient.setQueryDefaults(['todos'], { staleTime: 5000 })

useQuery(['todos'], fetchTodos)
queryClient.prefetchQuery(['todos'], fetchTodos)
```

both will have a `staleTime` of `5000`, and if you set a `queryFn` with `setQueryDefaults`, then both will have that as a default query fn. And of course, there you have the queryClient in the closure, because you need to invoke the function on the client itself:

```
queryClient.setQueryDefaults(
  ['todos'],
  {
    staleTime: 5000,
    queryFn: () => {
        const stuffSchema = await getSchema(queryClient, SCHEMA_STUFF)
    }
})

useQuery(['todos'])
queryClient.prefetchQuery(['todos'])
```

**Reply by GLabat** [reply]
> SSR stuff…

Sorry to have "polluted" the initial question… which was not from me…
Should we open a dedicated discussion?

---

> what I like to do is have sharable objects / factories for the queryKeys - even though I also have them as hooks sometimes (useQueryKeys()) because they internally use some id from the url for example. That is needed imo because we have many places where we need to have access to the key, mainly when performing query invalidation.

We definitely agree here. "Query Keys are key" and we should define the logic once. And provide helpers (hooks, factories…) to get a value without having to "know" how the key is made.
The same logic could be applied to invalidations as well. Define business-oriented helpers: `invalidateCompletedTodos` for ex to minimise key exposure.

---

> Another good way to share options between useQuery and prefetchQuery, if necessary, is to use queryClient.setQueryDefautls, which sets default options on a per key basis:

I think this is a good candidate for our sharing case, which still raises some questions.

Ex: 

- for `['todos']` key, the queryFn logic would be: get the schema, process schema, fetch todos
- for `['todos', someTodoId]` key, the queryFn logic would be: get the schema, process schema, fetch the given todo

So we would still have the need to factorise common part of the query function.

Will give it a try and get back to you.

---

About the addition of `queryClient` into `QueryFunctionContext`. Do you want us to create an issue?

Thanks.

**Reply by TkDodo** [reply]
> About the addition of queryClient into QueryFunctionContext. Do you want us to create an issue?

yes, let's track that in an issue please. I still need to think about it a bit, but right now, I could see this make it to v4 :)

**Reply by GLabat** [reply]
👌will do it tomorrow.

**Reply by GLabat** [reply]
Here we are: https://github.com/tannerlinsley/react-query/issues/3151

**Reply by GLabat** [reply]
I setup a code sandbox to test the queryDefaults pattern. It seems to work as expected and solve our main issue: having access to the queryClient inside a queryFn, outside of hook.

The sandbox is here: https://codesandbox.io/s/react-query-and-query-defaults-vywtm?file=/src/App.js

Would be great to have your feedback on this design.
Thanks.
